### PR TITLE
Add configurable keepAlive interval to detect half-open connections

### PR DESCRIPTION
**Pull request by bainian12345**
Send pong at a keepAliveInterval to ensure the control connection is not half-open

### CODE REVIEW

File: hyco-https/lib/HybridConnectionHttpsServer.js [hunk 1]
```diff
@@ -376,20 +375,42 @@ function connectControlChannel(server) {
...
+    var keepAliveInterval = null;
+    try {
+      keepAliveInterval = server.options.keepAliveTimeout.asMilliseconds();
+    } catch (ex) {
+      console.log("keepAliveTimeout should be an instance of moment.duration");
+    }
+    if (keepAliveInterval) {
+      keepAliveTimer = setInterval(function() {
+        try {
+          server.controlChannel.pong();
```
Comment (dlstucki) [inline hyco-https/lib/HybridConnectionHttpsServer.js]: >pong [](start = 32, length = 4)

Wow. I thought maybe only ping would be in the API.

File: hyco-https/lib/HybridConnectionHttpsServer.js [hunk 2]
```diff
@@ -376,20 +375,42 @@ function connectControlChannel(server) {
...
+        } catch (e) {
+          if (!server.closeRequested) {
+            // Closed unexpectedly, should reconnect
+            connectControlChannel(server);
+          }
```
Comment (dlstucki) [inline hyco-https/lib/HybridConnectionHttpsServer.js]: Should clearIntervals be called here?

Comment (bainian12345) [reply hyco-https/lib/HybridConnectionHttpsServer.js]: Yes that's correct

---
In reply to: [315839091](https://github.com/Azure/azure-relay-node/pull/34#discussion_r315839091) [](ancestors = 315839091)

**Comment by bainian12345** [comment]
>       connectControlChannel(server);

Reconnect needs a delay, will do it in a separate PR

---
Refers to: hyco-https/lib/HybridConnectionHttpsServer.js:418 in a1c19bf. [](commit_id = a1c19bfff9495e18a3c07f5ea6b3bde1400d5d8f, deletion_comment = False)

**Comment by dlstucki** [comment]
>       connectControlChannel(server);

delay after a failed ***re*** connect. The first re-connect attempt after being connected should not have any delay.

---
In reply to: [523198907](https://github.com/Azure/azure-relay-node/pull/34#issuecomment-523198907) [](ancestors = 523198907)

---
Refers to: hyco-https/lib/HybridConnectionHttpsServer.js:418 in a1c19bf. [](commit_id = a1c19bfff9495e18a3c07f5ea6b3bde1400d5d8f, deletion_comment = False)

**Pull request merged**